### PR TITLE
Blocked Access model, delegate and db migration changes for AB#15405

### PR DESCRIPTION
**Pull request by BrianMaki**
# Implements [AB#15405](https://dev.azure.com/qslvic/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/15405)

## Description

- Created migration to create BlockedAccess table
- Updated migration to rename DependentAudit to AgentAudit - includes column names
- Updated migration (Up and Down) to backup existing DependentAudit data to be inserted into AgentAudit after creation
- Updated migration to rename DependentOperationCode to OperationCode
- Renamed enum DependentAuditOperation to AuditOperation 
- Created BlockedAccess delegate to get BlockedAccess, add/update BlockedAccess and Delete BlockedAccess - this includes saves to AgentAudit
- Updated DbDelegationDelegate to query AgentAudit by type i.e., Dependent or ChangeDataSourceAccess
- Updated service code to reflect refactored code
- Updated unit tests to reflect refactored code
- Updated seed data script to reflect refactored code
- Added DataSource enum to be used for JsonB column in BlockedAccess
- Added GroupCode to AgentAudit using AuditGroupCode (AuditGroup enum) so that querying AgentAudit could be simplified as per suggestion from @ytqsl 

**DB Migration Up:**

Screenshot 2023-05-06 at 11 02 04 AM

Screenshot 2023-05-06 at 11 02 20 AM

**DB Migration Down:**

Screenshot 2023-05-06 at 11 01 05 AM

**Database:**

Screenshot 2023-05-05 at 4 35 53 PM

Screenshot 2023-05-05 at 4 37 08 PM

**Unit Tests:**

Screenshot 2023-05-05 at 1 35 24 PM

**Functional Tests:**

Screenshot 2023-05-05 at 1 16 52 PM

## Testing

- [x] Unit Tests Updated
- [ ] Functional Tests Updated
- [ ] Not Required

## UI Changes

None

## Items to Review:

-   [General PR Guidelines](https://github.com/bcgov/healthgateway/wiki/PRguidance)

### CODE REVIEW

File: Apps/Database/src/Delegates/IDelegationDelegate.cs [hunk 1]
```diff
@@ -36,19 +38,47 @@ public interface IDelegationDelegate
...
+        /// <summary>
+        /// Gets the hdid to search by.
+        /// </summary>
+        public string Hdid { get; init; } = string.Empty;
+    }
+
+    /// <summary>
+    /// Returns agent audit query results.
+    /// </summary>
+    public record AgentAuditQueryResult
```
Comment (BronzBierd) [inline Apps/Database/src/Delegates/IDelegationDelegate.cs:77]: This type of wrapper really isn't adding benefits, we wrap in delegate then immediately unwrap in services, just to put it in a new wrapper for delegation info. its just adding complexity.

**Review by BronzBierd — APPROVED** [review]

**Review by BronzBierd — CHANGES_REQUESTED** [review]
See previous review comments.

File: Apps/Database/src/Delegates/IDelegationDelegate.cs (hunk 1 above)
Comment (BrianMaki) [reply Apps/Database/src/Delegates/IDelegationDelegate.cs:77]: This is following the pattern that @ytqsl introduced in DbResourceDelegateDelegate.  Using a 'Query' and a 'Result' object was to ensure that any future changes to this query would not affect an existing service call.  Comments? @MikeLyttle  @rayking29  @sslaws  @ytqsl

File: Apps/Database/src/Models/AgentAudit.cs [hunk 2]
```diff
@@ -49,19 +49,26 @@ public class DependentAudit : AuditableEntity
...
         [Required]
         [MaxLength(10)]
-        public DependentAuditOperation OperationCode { get; set; } = DependentAuditOperation.Unprotect;
+        public AuditOperation OperationCode { get; set; } = AuditOperation.UnprotectDependent;
+
+        /// <summary>
+        /// Gets or sets the group.
+        /// </summary>
+        [Required]
+        [MaxLength(10)]
```
Comment (ytqsl) [inline Apps/Database/src/Models/AgentAudit.cs:70]: I think this need to be more than 10 - go with at least 50.

Comment (BrianMaki) [reply Apps/Database/src/Models/AgentAudit.cs:70]: @ytqsl A bug was going to be created to change all code tables regarding column length of that column.

Comment (ytqsl) [reply Apps/Database/src/Models/AgentAudit.cs:70]: As this is a new column, wouldn't it be better to increase the length before merging to dev?

**Review by ytqsl — APPROVED** [review]

**Review by BronzBierd — APPROVED** [review]

File: Apps/Database/src/Models/AgentAudit.cs (hunk 2 above)
Comment (BrianMaki) [reply Apps/Database/src/Models/AgentAudit.cs]: Changed.

**Pull request merged**